ldms/src/sampler/spank/slurm_notifier.c

```diff
@@ -74,9 +74,11 @@
 static char *stream;
 #define SLURM_NOTIFY_TIMEOUT 5
 static time_t io_timeout = SLURM_NOTIFY_TIMEOUT;
+static int user_debug = 0;
 
 static const char *stepd_event = "";
 #define DEBUG2(FMT, ...) do { \
+	if (user_debug) \
```

Comment:
It seems like this is left-over code from when the sampler was being developed. I seem to recall that originally this code called slurm_debug2(...), not printf directly and that the slurm configuration debug level controlled what was seen. In any event it seems like that is the way that it should work, not with a global variable like this. BTW this is consistent with the spank plugin documentation here: https://slurm.schedmd.com/plugins.html

Reply:
While I agree the obvious thing to do would be to revert Mon's redefinition of DEBUG2 that used printf, it turns out that from an admin-controlled feature point of view it is much easier to debug/trace what the new plugin is doing by examining the user log file with the spank messages in context. As a general rule in my shop, the slurm daemon logs get routed off to splunk and are difficult or impossible to access for the performance monitoring team. 

I'd be happy to update the patch to provide 

- 'debug=0' no logging
- 'debug=1' log to user log
- 'debug=2' log to slurm_debug2

Thoughts @nichamon ?

Reply:
I'm not entirely sure that all of slurm_* log functions are available, are they? There seems to be a slurm_spank_log(), but that is for printing messages back through srun, I think. I really don't know.

Reply:
> 

The logging functions are all defined in spank.h and I have tested that they are in fact available to plugins. The macro in question DEBUG2 used to call slurm_debug2 instead of printf; the change was made (and accidentally left in) when debugging the plugin because as Ben points out the slurm_debug2 goes to the slurm log instead of the output of the job.

It is more convenient when debugging to see these messages in the job output then looking at the slurm logs. That said, IMO the default behavior should be that plugin messages go to the slurm log as governed by the slurm DebugLevel configuration, not a plugin specific configuration. As long as the default behavior is to let the slurm configuration govern where and what messages are logged, then I am good with this.

It should be noted that per Ben's change when in 'debug' mode, _all_ messages end up in the job output and the configured DebugLevel is ignored. That said, when in debug mode ... who cares...

From spank.h 
```
/*
 *  Slurm logging functions which are exported to plugins.
 */
extern void slurm_info (const char *format, ...)
  __attribute__ ((format (printf, 1, 2)));
extern void slurm_error (const char *format, ...)
  __attribute__ ((format (printf, 1, 2)));
extern void slurm_verbose (const char *format, ...)
  __attribute__ ((format (printf, 1, 2)));
extern void slurm_debug (const char *format, ...)
  __attribute__ ((format (printf, 1, 2)));
extern void slurm_debug2 (const char *format, ...)
  __attribute__ ((format (printf, 1, 2)));
extern void slurm_debug3 (const char *format, ...)
  __attribute__ ((format (printf, 1, 2)));

/*
 * Print at the same log level as error(), but without prefixing the message
 * with "error: ". Useful to report back to srun commands from SPANK plugins,
 * as info() will only go to the logs.
 */
extern void slurm_spank_log(const char *, ...)
  __attribute__ ((format (printf, 1, 2)));
```

Reply:
Whether or not they actually go to the slurm log might depends on which context the spank plugin is called from. I'm not sure that all will result in them going there.

Reply:
@morrone, right ... sometimes these log functions are called from slurmd, sometimes from slurmstepd (srun/root) and sometimes from slurmstepd (user). But I think slurm takes care of that ... at least that is what I've seen in the past.